primitives/src/v1.rs

```diff
@@ -705,6 +708,37 @@ sp_api::decl_runtime_apis! {
 	}
 }
 
+/// The output of a collator.
+#[derive(Clone, Encode, Decode)]
+pub struct Collation {
+	/// Head data of the para block.
+	pub head_data: HeadData,
+	/// Messages to be passed up to the relay chain.
+	pub upward_messages: Vec<UpwardMessage>,
+	/// Proof that this block is valid.
+	pub proof_of_validity: PoV,
+}
+
+/// Configuration for the collation generator
+#[cfg(feature = "std")]
+pub struct CollationGenerationConfig {
+	/// Collator's authentication key, so it can sign things.
+	pub key: CollatorPair,
+	/// Collation function.
+	// REVIEW: the guide memtions generating the validation function params here, but that struct
+	// is only defined within cumulus, and I don't want to either add a circular dependency or
+	// move the struct definition into here. Would there be any real pitfalls to just providing both
+	// pieces of validation data and letting the collator extract its own params?
+	pub collator: Box<dyn Fn(&GlobalValidationData, &LocalValidationData) -> Box<dyn Future<Output = Collation> + Unpin + Send> + Send + Sync>,
```

Comment:
I don't see any reason not to just hand these functions to the collator, but maybe there's one I didn't consider.

Reply:
As an aside: This should be in `node-primitives`, not in `polkadot-primitives`, since this never will hit the runtime.

I think passing the validation data into this function is fine and this is probably what the guide text means - either that or the `ValidationParams` from `polkadot-parachain`. But actually #1539 makes a strong case that the collator logic will want all the validation data, not just the direct parameters to Wasm, so it can construct a block under the non-persisted constraints.